precompiles/staking/logs.go

```diff
@@ -0,0 +1,154 @@
+package staking
+
+import (
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+func (c *Contract) AddStakeLog(
+	ctx sdk.Context,
+	stateDB vm.StateDB,
+	staker common.Address,
+	validator string,
+	amount *big.Int,
+) error {
+	event := c.Abi().Events["Stake"]
+
+	valAddr, err := sdk.ValAddressFromBech32(validator)
+	if err != nil {
+		return err
+	}
+
+	// staker and validator are indexed event params
+	topics, err := makeTopics(event, []interface{}{staker}, []interface{}{common.BytesToAddress(valAddr.Bytes())})
+	if err != nil {
+		return err
+	}
+
+	// amount is part of event data
+	data, err := packAmount(amount)
+	if err != nil {
+		return err
+	}
+
+	c.addLog(ctx, stateDB, topics, data)
+
+	return nil
+}
+
+func (c *Contract) AddUnstakeLog(
+	ctx sdk.Context,
+	stateDB vm.StateDB,
+	staker common.Address,
+	validator string,
+	amount *big.Int,
+) error {
+	event := c.Abi().Events["Unstake"]
+	valAddr, err := sdk.ValAddressFromBech32(validator)
+	if err != nil {
+		return err
+	}
+
+	// staker and validator are indexed event params
+	topics, err := makeTopics(event, []interface{}{staker}, []interface{}{common.BytesToAddress(valAddr.Bytes())})
+	if err != nil {
+		return err
+	}
+
+	// amount is part of event data
+	data, err := packAmount(amount)
+	if err != nil {
+		return err
+	}
+
+	c.addLog(ctx, stateDB, topics, data)
+
+	return nil
+}
+
+func (c *Contract) AddMoveStakeLog(
+	ctx sdk.Context,
+	stateDB vm.StateDB,
+	staker common.Address,
+	validatorSrc string,
+	validatorDst string,
+	amount *big.Int,
+) error {
+	event := c.Abi().Events["MoveStake"]
+	validatorSrcAddr, err := sdk.ValAddressFromBech32(validatorSrc)
+	if err != nil {
+		return err
+	}
+
+	validatorDstAddr, err := sdk.ValAddressFromBech32(validatorDst)
+	if err != nil {
+		return err
+	}
+
+	// staker and validators are indexed event params
+	topics, err := makeTopics(
+		event,
+		[]interface{}{staker},
+		[]interface{}{common.BytesToAddress(validatorSrcAddr.Bytes())},
+		[]interface{}{common.BytesToAddress(validatorDstAddr.Bytes())},
+	)
+	if err != nil {
+		return err
+	}
+
+	// amount is part of event data
+	data, err := packAmount(amount)
+	if err != nil {
+		return err
+	}
+
+	c.addLog(ctx, stateDB, topics, data)
+
+	return nil
+}
+
+func (c *Contract) addLog(ctx sdk.Context, stateDB vm.StateDB, topics []common.Hash, data []byte) {
+	stateDB.AddLog(&types.Log{
+		Address:     c.Address(),
+		Topics:      topics,
+		Data:        data,
+		BlockNumber: uint64(ctx.BlockHeight()),
+	})
+}
```

Comment:
This function can be moved to precompiles package and be used from any contract.

Reply:
Maybe most function in this package could be made reusable, I think many contract will use packAmount as well